rfcs/0078-configuration-file-generators.md

```diff
@@ -0,0 +1,146 @@
+---
+feature: configuration_file_generators
+start-date: 2020-10-18
+author: Michael Raskin @7c6f434c
+co-authors: (in-progress)
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Provide the configuration file generation functionality separately from NixOS
+as a whole, and with scoped use of the module system.
+
+# Motivation
+[motivation]: #motivation
+
+There is currently a lot of code duplication between
+* NixOS
+* home-manager
+* nix-darwin
+
+and obviously growing code duplication with any effort running on below-Tier-2
+platforms. That leads both to effort duplication, and to confusion among users
+using multiple Nix ecosystem tools for service generation,
+
+While fully resolving the problem is likely not fully compatible with the core
+preferences of maintainers on each specific platform, we believe that config
+file generation could be a shared resource just like Nixpkgs based on the
+following properties:
+* unlike, say, upstream `systemd` unit files we typically do not
+  reuse-and-override upstream configs, but generate our own (both in NixOS and
+  other service collections)
+* the configs for a piece of software often have few dependencies on the
+  general environment
+* it can be defined as a purely functional library, simplifying integration
+
+We aim to reduce duplication of effort and code while also having a high level
+of inspection and tweaking possibilities without needing to patch module
+source.
+
+# Detailed design
+[design]: #detailed-design
+
+This RFC proposal builds upon RFCs#42 defining a configuration file abstraction
+approach.
+
+A subtree is established in the Nixpkgs repository serving as a library of
+functions implementing program configuration file generation. .As the design
+uses a module system, putting configuration generators next to packages is also
+not a perfect solution; also a configuration generator might end up generating
+a set of config for a combination of packages.
+
+Each such generator takes as an input a NixOS module system based «subtree»,
+e.g. the values typically bound to `cfg` variable in typical modules.
+
+The output of each such function is an attribute set. The keys are relative
+file names of configuration files, and the values are attribute sets of RFCs#42
+settings abstraction and serialiser.
+
+Along with the function the corresponding type specifications for input and
+output modules are provided.
+
+In some cases we only provide low-level overridable default configuration. In
+this case the input may have the type that is always an empty attribute set,
+`{}`.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+A minimalistic silly example is:
+
+Input:
+```
+  {
+     port = 1234;
+     nodeName = "localhost";
+     content = "Hello";
```

Comment:
Is this typed?

Reply:
Yes, the type is provided next to the generator function.